File: 213-house-robber-ii/213-house-robber-ii.cpp
class Solution {
    vector<int> dp1, dp2;
public:
    int rob(vector<int>& nums) {
        int n=nums.size();
        if (n==1) return nums[0];
        dp1.assign(n+1, -1);
        dp2.assign(n+1, -1);
        return max(fun(nums, 0, n-2, dp1), fun(nums, 1,n-1, dp2));
    }
private:
    int fun(vector<int>& nums, int si, int ei, vector<int> &dp) {
        if (si > ei) return 0;
        if (dp[si] != -1) 
            return dp[si];
        return dp[si] = max(nums[si]+fun(nums, si+2, ei, dp), fun(nums, si+1, ei, dp));
    }
};

File: 213-house-robber-ii/213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def HouseRobber1(si, ei):
            #[rob1, rob2, ele1, ele2]
            if len(nums) <= 2:
                return max(nums[0], nums[1]) if len(nums) == 2 else nums[0]
            rob1,rob2 = 0, 0
            for i in range(si, ei+1):
                cur = max(nums[i]+ rob1, rob2)
                rob1 = rob2
                rob2 = cur
            return rob2
        return max(HouseRobber1(0, len(nums)-2), HouseRobber1(1, len(nums)-1))